### PR TITLE
BUG: Bug in loc raised ValueError when setting value via boolean list

### LINKED ISSUE #20438: iloc boolean mask setitem
Is this supported?

```python
x = pd.Series([0, 1, 2])
x.iloc[[True, False, True]] = [1, 2]
```

Currently that raises with

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-aa3d14fe6683> in <module>()
      1 x = pd.Series([0, 1, 2])
----> 2 x.iloc[[True, False, True]] = [1, 2]

~/sandbox/pandas-ip/pandas/pandas/core/indexing.py in __setitem__(self, key, value)
    185             key = com._apply_if_callable(key, self.obj)
    186         indexer = self._get_setitem_indexer(key)
--> 187         self._setitem_with_indexer(indexer, value)
    188
    189     def _has_valid_type(self, k, axis):

~/sandbox/pandas-ip/pandas/pandas/core/indexing.py in _setitem_with_indexer(self, indexer, value)
    636             self.obj._consolidate_inplace()
    637             self.obj._data = self.obj._data.setitem(indexer=indexer,
--> 638                                                     value=value)
    639             self.obj._maybe_update_cacher(clear=True)
    640

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in setitem(self, **kwargs)
   3646
   3647     def setitem(self, **kwargs):
-> 3648         return self.apply('setitem', **kwargs)
   3649
   3650     def putmask(self, **kwargs):

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in apply(self, f, axes, filter, do_integrity_check, consolidate, **kwargs)
   3534
   3535             kwargs['mgr'] = self
-> 3536             applied = getattr(b, f)(**kwargs)
   3537             result_blocks = _extend_blocks(applied, result_blocks)
   3538

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in setitem(self, indexer, value, mgr)
    883                         indexer.dtype == np.bool_ and
    884                         len(indexer[indexer]) == len(value)):
--> 885                     raise ValueError("cannot set using a list-like indexer "
    886                                      "with a different length than the value")
    887

ValueError: cannot set using a list-like indexer with a different length than the value
```

But `__setitem__` and `.loc` support setting via that.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #20438
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Not quite sure if conversion to numpy array is necessary, but ``is_dtype_bool`` does not accept lists

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -602,7 +603,7 @@ def _get_setitem_indexer(self, key):
         """
         Convert a potentially-label-based key into a positional indexer.
         """
-        if self.name == "loc":
+        if self.name == "loc" or is_bool_dtype(np.array(key)):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:606]: is it only lists we want here, or would `com.is_bool_indexer(key)` work?

File: pandas/tests/indexing/test_loc.py [hunk 2]
```diff
@@ -1955,6 +1955,13 @@ def test_loc_setitem_dt64tz_values(self):
         result = s2["a"]
         assert result == expected
 
+    def test_loc_setitem_boolean_list(self):
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_loc.py:1958]: lets put this in TestLocBooleanMask

can this be shared by DataFrame?

few words after the GH ref to the effect of "specifically list key, not arraylike"

File: pandas/core/indexing.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexing.py:606]: That is better, changed it :)

File: pandas/tests/indexing/test_loc.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py]: Done, added a ``DataFrame`` part to the test, parametrization would be quite ugly here

File: pandas/tests/indexing/test_loc.py [hunk 3]
```diff
@@ -1644,6 +1644,18 @@ def test_loc_setitem_mask_td64_series_value(self):
         assert expected == result
         tm.assert_frame_equal(df, df_copy)
 
+    def test_loc_setitem_boolean_list(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py:1647]: can you parameterize the rhs by array as well as list

Comment (jreback) [reply pandas/tests/indexing/test_loc.py:1647]: ahh i see the comment, but still think can parameterize.

**Review by jreback — CHANGES_REQUESTED** [review]
also the original issue wa about iloc, can you test that as well (not sure if this is a *different* bug then)

File: pandas/tests/indexing/test_loc.py (hunk 3 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py]: Done

**Comment by phofl** [comment]
cc @jreback Thanks very much, iloc actually was a different issue. 

The validation in 
https://github.com/pandas-dev/pandas/blob/b96665777c083516c9cdf21d25b7d9ccf70c81bd/pandas/core/indexers.py#L153
did not allow boolean lists. We have two options to fix this for iloc. I converted a boolean list to an array in the indexing module, because I was not sure if this validation was done on purpose this way. Alternatively we could adjust the referenced validation to allow boolean lists too

File: pandas/tests/indexing/test_loc.py [hunk 4]
```diff
@@ -1645,6 +1645,20 @@ def test_loc_setitem_mask_td64_series_value(self):
         assert expected == result
         tm.assert_frame_equal(df, df_copy)
 
+    @pytest.mark.parametrize("func", ["loc", "iloc"])
+    @pytest.mark.parametrize("rhs_func", [list, np.array])
+    def test_loc_setitem_boolean_list(self, func, rhs_func):
+        # GH#20438 testing specifically list key, not arraylike
+        ser = Series([0, 1, 2])
+        getattr(ser, func)[[True, False, True]] = rhs_func([5, 10])
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py:1653]: can you *also* parameterize on the indexing type as well (e.g. list / array / pd.Index)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @jbrockmendel

File: pandas/tests/indexing/test_loc.py (hunk 4 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py]: Index is not allowed for iloc, added the remaining ones

File: doc/source/whatsnew/v1.2.0.rst [hunk 5]
```diff
@@ -476,6 +476,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc` returned requested key plus missing values when ``loc`` was applied to single level from :class:`MultiIndex` (:issue:`27104`)
 - Bug in indexing on a :class:`Series` or :class:`DataFrame` with a :class:`CategoricalIndex` using a listlike indexer containing NA values (:issue:`37722`)
 - Bug in :meth:`DataFrame.xs` ignored ``droplevel=False`` for columns (:issue:`19056`)
+- Bug in :meth:`Series.loc` raised an Error when input was filtered with a boolean list and values to set were a list with lower dimension (:issue:`20438`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:479]: is there a specific exception we can say instead of Error?

File: pandas/tests/indexing/test_loc.py [hunk 6]
```diff
@@ -1645,6 +1645,21 @@ def test_loc_setitem_mask_td64_series_value(self):
         assert expected == result
         tm.assert_frame_equal(df, df_copy)
 
+    @pytest.mark.parametrize("indexing_func", [list, np.array])
+    @pytest.mark.parametrize("func", ["loc", "iloc"])
+    @pytest.mark.parametrize("rhs_func", [list, np.array])
+    def test_loc_setitem_boolean_list(self, func, rhs_func, indexing_func):
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_loc.py:1651]: its only the *-iloc-list cases that fail on master.  at its heart, is this an iloc bug or a loc bug?

File: pandas/tests/indexing/test_loc.py [hunk 7]
```diff
@@ -1645,6 +1645,21 @@ def test_loc_setitem_mask_td64_series_value(self):
...
+        # GH#20438 testing specifically list key, not arraylike
+        ser = Series([0, 1, 2])
+        getattr(ser, func)[indexing_func([True, False, True])] = rhs_func([5, 10])
```
Comment (jbrockmendel) [inline pandas/tests/indexing/test_loc.py:1654]: in the np.array case, do we care if the ndarray has bool vs object dtype?

File: pandas/core/indexing.py [hunk 8]
```diff
@@ -602,7 +602,7 @@ def _get_setitem_indexer(self, key):
         """
         Convert a potentially-label-based key into a positional indexer.
         """
-        if self.name == "loc":
+        if self.name == "loc" or com.is_bool_indexer(key):
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:605]: i haven't fully formed an opinion, but i _think_ another alternative would be to properly handle boolean masks in Block.setitem.

**Comment by jbrockmendel** [comment]
does this have a perf impact?  IIRC is_bool_indexer isn't completely cheap

File: pandas/core/indexing.py (hunk 8 above)
Comment (phofl) [reply pandas/core/indexing.py:605]: You are right, loc is not failing, this means we do not need this or. Removed it.

File: pandas/tests/indexing/test_loc.py (hunk 6 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py:1651]: iloc, thx.

File: doc/source/whatsnew/v1.2.0.rst (hunk 5 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst:479]: ValueError, added it

File: pandas/tests/indexing/test_loc.py (hunk 7 above)
Comment (phofl) [reply pandas/tests/indexing/test_loc.py:1654]: Object dtype is killed by https://github.com/pandas-dev/pandas/blob/b96665777c083516c9cdf21d25b7d9ccf70c81bd/pandas/core/indexers.py#L153:L157 and raises a ValueError too.

File: pandas/core/indexing.py (hunk 8 above)
Comment (phofl) [reply pandas/core/indexing.py:605]: Lists can be handled in Block.setitem already, but arrays from dtype object don't work. Lists are killed by the validation mentioned above in ``pandas.core.indexers``.

**Comment by phofl** [comment]
Ran indexing asvs
before           after         ratio
     [50ae0bfc]       [b5dd4534]
     <master>         <20438>   
-         249±9ms          225±8ms     0.90  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-        30.5±2ms       27.4±0.4ms     0.90  indexing.ChainIndexing.time_chained_indexing(None)
-         252±7ms          218±8ms     0.87  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-         255±7ms          215±5ms     0.84  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-         181±3ms          144±2ms     0.80  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-         185±4ms          147±2ms     0.79  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-         190±9ms          150±3ms     0.79  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        43.1±1μs       26.1±0.9μs     0.60  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-     1.86±0.04ms      1.11±0.04ms     0.60  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-     1.70±0.05ms         953±30μs     0.56  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-         308±8μs          166±6μs     0.54  indexing.NumericSeriesIndexing.time_getitem_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        296±10μs          159±4μs     0.54  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        86.8±3μs         44.3±1μs     0.51  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: doc/source/whatsnew/v1.2.0.rst [hunk 9]
```diff
@@ -476,6 +476,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc` returned requested key plus missing values when ``loc`` was applied to single level from :class:`MultiIndex` (:issue:`27104`)
 - Bug in indexing on a :class:`Series` or :class:`DataFrame` with a :class:`CategoricalIndex` using a listlike indexer containing NA values (:issue:`37722`)
 - Bug in :meth:`DataFrame.xs` ignored ``droplevel=False`` for columns (:issue:`19056`)
+- Bug in :meth:`Series.loc` raised a ``ValueError`` when input was filtered with a boolean list and values to set were a list with lower dimension (:issue:`20438`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:479]: raised -> raising

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -1521,6 +1521,8 @@ def _convert_to_indexer(self, key, axis: int, is_setter: bool = False):
 
     def _get_setitem_indexer(self, key):
         # GH#32257 Fall through to let numpy do validation
+        if isinstance(key, list) and com.is_bool_indexer(key):
+            return np.array(key)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py:1525]: comment with `# GH#37761`

**Comment by jbrockmendel** [comment]
couple comments, needs rebase otherwise LGTM

File: doc/source/whatsnew/v1.2.0.rst (hunk 9 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst]: Thx

File: pandas/core/indexing.py (hunk 10 above)
Comment (phofl) [reply pandas/core/indexing.py:1526]: Added

**Comment by phofl** [comment]
ping @jbrockmendel ?

**Comment by jbrockmendel** [comment]
It looks like in master the test case raises in check_setitem_lengths.  would it make sense to patch at that point?

**Comment by phofl** [comment]
We could check for lists there too.

File: pandas/core/indexers.py [hunk 11]
```diff
@@ -152,6 +152,8 @@ def check_setitem_lengths(indexer, value, values) -> bool:
         if is_list_like(value):
             if len(indexer) != len(value):
                 # boolean with truth values == len of the value is ok too
+                if isinstance(indexer, list):
```
Comment (jreback) [inline pandas/core/indexers.py:155]: should't this be is_list_like? (e.g. if the indexer is a np.array?)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexers.py (hunk 11 above)
Comment (phofl) [reply pandas/core/indexers.py:155]: If we already have an array, we don't have to cast to array

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@phofl can you rebase (i am not sure when it was before), and ping on green.

**Comment by phofl** [comment]
@jreback greenish, we have same number of failures as on master

**Comment by simonjayhawkins** [comment]
Thanks @phofl

**Pull request merged**